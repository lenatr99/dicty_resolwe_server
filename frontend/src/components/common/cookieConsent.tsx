import { useTheme } from '@mui/material';
import React, { ReactElement } from 'react';
import ReactCookieConsent from 'react-cookie-consent';

const CookieConsent = (): ReactElement => {
    const theme = useTheme();

    return (
        <ReactCookieConsent
            buttonStyle={{
                background: theme.palette.secondary.main,
                color: 'white',
                borderRadius: theme.shape.borderRadius,
            }}
        >
            This website uses cookies to ensure you get the best experience on our website.{' '}
            <a
                href="https://www.genialis.com/privacy-policy/"
                rel="noopener noreferrer"
                target="_blank"
            >
                Learn more
            </a>
        </ReactCookieConsent>
    );
};

export default CookieConsent;
