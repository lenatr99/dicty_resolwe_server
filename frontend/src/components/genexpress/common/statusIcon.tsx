import type * as TIcons from '@mui/icons-material';
import * as Icons from '@mui/icons-material';
import {
    DataStatus,
    DIRTY_DATA_STATUS,
    DONE_DATA_STATUS,
    ERROR_DATA_STATUS,
    PREPARING_DATA_STATUS,
    PROCESSING_DATA_STATUS,
    RESOLVING_DATA_STATUS,
    UPLOADING_DATA_STATUS,
    WAITING_DATA_STATUS,
} from '@genialis/resolwe/dist/api/types/rest';
import { Tooltip } from '@mui/material';

export const StatusIconMap = {
    [UPLOADING_DATA_STATUS]: 'UploadFile',
    [RESOLVING_DATA_STATUS]: 'Schedule',
    [PREPARING_DATA_STATUS]: 'Archive',
    [WAITING_DATA_STATUS]: 'Schedule',
    [PROCESSING_DATA_STATUS]: 'DataUsage',
    [DONE_DATA_STATUS]: 'DoneAll',
    [ERROR_DATA_STATUS]: 'Error',
    [DIRTY_DATA_STATUS]: 'BugReport',
};

export const StatusLabelMap = {
    [UPLOADING_DATA_STATUS]: 'Uploading',
    [RESOLVING_DATA_STATUS]: 'Waiting in queue',
    [PREPARING_DATA_STATUS]: 'Preparing',
    [WAITING_DATA_STATUS]: 'Waiting in queue',
    [PROCESSING_DATA_STATUS]: 'Processing',
    [DONE_DATA_STATUS]: 'Done',
    [ERROR_DATA_STATUS]: 'Error',
    [DIRTY_DATA_STATUS]: 'Dirty',
};

export const StatusIcon = ({ status }: { status: DataStatus }) => {
    const IconComponentName = StatusIconMap[status];
    const Icon = Icons[IconComponentName as keyof typeof TIcons];

    if (status === DONE_DATA_STATUS) {
        return null;
    }

    return (
        <Tooltip title={StatusLabelMap[status]}>
            <Icon fontSize="small" />
        </Tooltip>
    );
};
