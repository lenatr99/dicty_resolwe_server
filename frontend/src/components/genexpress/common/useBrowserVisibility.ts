import { useEffect } from 'react';

export const useBrowserVisibility = ({
    onShow,
    onHide,
}: {
    onShow?: () => void;
    onHide?: () => void;
}) => {
    useEffect(() => {
        const handleVisibilityChange = () => {
            const visibilityState = document.visibilityState;
            if (visibilityState === 'visible') {
                onShow?.();
            } else if (visibilityState === 'hidden') {
                onHide?.();
            }
        };

        document.addEventListener('visibilitychange', handleVisibilityChange);

        return () => {
            document.removeEventListener('visibilitychange', handleVisibilityChange);
        };
    }, [onShow, onHide]);
};

export default useBrowserVisibility;
