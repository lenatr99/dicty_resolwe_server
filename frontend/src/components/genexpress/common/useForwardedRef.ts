import { MutableRefObject, useEffect, useRef } from 'react';

/**
 * If ref of an underlying element has to be exposed, use ref from this hook to handle
 * cases when parent component doesn't need the ref..
 * @param ref ForwardedRef. If it's null, a new ref object is created.
 */
const useForwardedRef = <T>(
    ref: ((instance: T | null) => void) | React.MutableRefObject<T | null> | null | undefined,
): MutableRefObject<T | null> => {
    const innerRef = useRef<T>(null);
    useEffect(() => {
        if (!ref) return;
        if (typeof ref === 'function') {
            ref(innerRef.current);
        } else {
            ref.current = innerRef.current;
        }
    });

    return innerRef;
};

export default useForwardedRef;
