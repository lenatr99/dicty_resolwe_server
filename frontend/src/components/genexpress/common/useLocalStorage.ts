import { useEffect, useState, useCallback } from 'react';
import {
    getLocalStorageKey,
    writeToLocalStorage,
    readFromLocalStorage,
    deleteFromLocalStorage,
} from '../../../utils/localStorageUtils';

/**
 * React hook to enable updates to state via localStorage.
 * This updates when the {writeStorage} function is used, when the returned function
 * is called.
 * This function takes an optional default value to start off with.
 *
 * @example
 * ```js
 * const MyComponent = () => {
 *   const [myStoredItem, setMyStoredItem] = useLocalStorage('myStoredItem', 'initialValue');
 *   return (
 *     <p>{myStoredItem}</p>
 *   );
 * };
 * ```
 *
 * @export
 * @template TValue The type of the given initial value.
 * @param {string} targetKey The key in the localStorage that you wish to watch.
 * @param {TValue} initialValue Initial value to start with.
 * @returns {[TValue, Dispatch<TValue>, Dispatch<void>]} An array containing the value
 * associated with the key in position 0, a function to set the value in position 1,
 * and a function to delete the value from localStorage in position 2.
 */
const useLocalStorage = <TValue = string>(
    targetKey: string,
    initialValue: TValue,
): [TValue, (value: TValue | ((prevState: TValue) => TValue)) => void, () => void] => {
    const key = getLocalStorageKey(targetKey);

    const [localState, updateLocalState] = useState<TValue>(
        readFromLocalStorage(key) ?? initialValue,
    );

    // Each time local state changes, write it's new value to local storage.
    useEffect(() => {
        writeToLocalStorage(key, localState);
    }, [key, localState]);

    useEffect(() => {
        const itemExists = readFromLocalStorage(key) != null;

        // Write initial value to the local storage if it's not present or contains invalid JSON data.
        if (initialValue !== undefined && !itemExists) {
            writeToLocalStorage(key, initialValue);
        }
    }, [initialValue, key]);

    const deleteState = useCallback(() => {
        updateLocalState(initialValue);
        deleteFromLocalStorage(key);
    }, [initialValue, key]);

    return [localState === null ? initialValue : localState, updateLocalState, deleteState];
};

export default useLocalStorage;
