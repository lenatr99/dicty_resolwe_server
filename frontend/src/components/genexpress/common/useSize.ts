import useResizeObserver, { UseResizeObserverCallback } from '@react-hook/resize-observer';
import { debounce, pick } from 'lodash';
import { RefObject, useEffect, useLayoutEffect, useRef, useState } from 'react';

const useSize = (targetRef: RefObject<HTMLElement>, refreshRate = 16) => {
    const [size, setSize] = useState<{ width: number; height: number }>({ width: 0, height: 0 });
    const isUnmounted = useRef(false);
    const debouncedSetSize = useRef(
        debounce<UseResizeObserverCallback>((entry) => {
            if (!isUnmounted.current) {
                setSize(entry.contentRect);
            }
        }, refreshRate),
    );

    useLayoutEffect(() => {
        if (targetRef.current != null) {
            setSize(pick(targetRef.current?.getBoundingClientRect(), ['width', 'height']));
        }
    }, [targetRef]);

    const resizeObserverRef = useResizeObserver(targetRef, debouncedSetSize.current);

    useEffect(() => {
        isUnmounted.current = false;
        const targetRefCurrent = targetRef.current;
        const debouncedSetSizeCurrent = debouncedSetSize.current;

        return () => {
            if (targetRefCurrent != null) {
                resizeObserverRef.unobserve(targetRefCurrent);
            }
            debouncedSetSizeCurrent.cancel();
            isUnmounted.current = true;
        };
    }, [resizeObserverRef, targetRef]);

    return size;
};

export default useSize;
