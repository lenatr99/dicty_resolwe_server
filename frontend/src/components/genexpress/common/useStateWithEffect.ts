import { useEffect, useState } from 'react';

const useStateWithEffect = <T>(getValue: () => T, dependencies: unknown[]): T => {
    const [value, setValue] = useState<T>(getValue);

    useEffect(() => {
        setValue(getValue());
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, dependencies);

    return value;
};

export default useStateWithEffect;
