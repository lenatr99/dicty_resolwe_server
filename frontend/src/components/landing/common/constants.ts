enum SectionNames {
    FEATURES = 'features',
    SCREENSHOTS = 'screenshots',
    CITING = 'citing',
    REFERENCES = 'references',
    CREDITS = 'credits',
    CONTACT = 'contact',
}

export default SectionNames;
