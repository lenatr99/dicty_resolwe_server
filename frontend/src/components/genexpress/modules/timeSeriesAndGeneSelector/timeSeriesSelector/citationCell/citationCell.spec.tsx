import { fireEvent, screen } from '@testing-library/react';
import { Relation } from '@genialis/resolwe/dist/api/types/rest';
import CitationCell from './citationCell';
import { customRender } from 'tests/test-utils';
import { generateSingleTimeSeries } from 'tests/mock';
import { DescriptorSchemaSlug } from 'components/genexpress/common/constants';
import { Descriptor } from 'types/application';

const relation = generateSingleTimeSeries(1) as Relation & {
    descriptor: Required<Descriptor[DescriptorSchemaSlug.DictyTimeSeries]>;
};

describe('citationCell', () => {
    let asFragment: () => DocumentFragment;

    beforeEach(() => {
        ({ asFragment } = customRender(<CitationCell data={relation} />));
    });

    it('should render as default snapshot', () => {
        expect(asFragment()).toMatchSnapshot();
    });

    it('should open citation dialog on click', () => {
        fireEvent.click(screen.getByRole('button', { name: relation.descriptor.citation.name }));

        screen.getByText(
            'dictyExpress: a web-based platform for sequence data management and analytics in Dictyostelium and beyond',
            { exact: false },
        );
        screen.getByText('2017', { exact: false });
        screen.getByText('BMC Bioinformatics', { exact: false });
        expect(
            screen.getByRole('link', { name: relation.descriptor.citation.name }),
        ).toHaveAttribute('href', relation.descriptor.citation.url);
        screen.getByRole('heading', {
            name: `${relation.descriptor.project} ${relation.descriptor.details}`,
        });
    });
});
