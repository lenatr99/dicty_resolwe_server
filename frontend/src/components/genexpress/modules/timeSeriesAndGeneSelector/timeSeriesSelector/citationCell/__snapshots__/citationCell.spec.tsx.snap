// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`citationCell > should render as default snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gJhJTp fonkkw MuiBox-root"
  >
    <button
      class="MuiButtonBase-root-ocQuz kiSjtm MuiButtonBase-root"
      tabindex="0"
      type="button"
    >
      <a
        class="MuiTypography-root-hJQhZH cagcOC MuiTypography-root MuiTypography-inherit MuiLink-root-efECBU drMBtS MuiLink-root MuiLink-underlineAlways"
      >
        Rosengarten et. al.
      </a>
      <span
        class="MuiTouchRipple-root-huKFMv dxeyKA MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
